tests/gadget-zero/test_gadget0.py

```diff
@@ -401,5 +410,73 @@ def test_unaligned(self):
 
 
+class TestControlTransfer_Delays(unittest.TestCase):
+    """
+    Test cases for
+    https://github.com/libopencm3/libopencm3/issue/668
+    """
+
+    def setUp(self):
+        self.dev = usb.core.find(idVendor=VENDOR_ID, idProduct=PRODUCT_ID, custom_match=find_by_serial(DUT_SERIAL))
+        self.assertIsNotNone(self.dev, "Couldn't find locm3 gadget0 device")
+
+        self.cfg = uu.find_descriptor(self.dev, bConfigurationValue=4)
+        self.assertIsNotNone(self.cfg, "Config 4 should exist")
+        self.dev.set_configuration(self.cfg)
+        self.req     = uu.CTRL_IN  | uu.CTRL_TYPE_VENDOR | uu.CTRL_RECIPIENT_INTERFACE
+        self.req_out = uu.CTRL_OUT | uu.CTRL_TYPE_VENDOR | uu.CTRL_RECIPIENT_INTERFACE
+        self.intr_out = 1
+        # send dummy packages on interrupt port to get in sync
+        # (alternating bit protocol)
+        # FIXME: Why is this necessary?
+        cmd = struct.pack('<B', GZ_INTR_CMD_UNDEFINED)
+        self.dev.write(self.intr_out, cmd)
+        self.dev.write(self.intr_out, cmd)
+
+    def tearDown(self):
+        uu.dispose_resources(self.dev)
+
+    def read_string(self, idx, lang=0x0409):
+        req = uu.CTRL_IN | uu.CTRL_TYPE_STANDARD | uu.CTRL_RECIPIENT_DEVICE
+        return self.dev.ctrl_transfer(req, uu.GET_DESCRIPTOR, idx, lang, 255)
+
+    def test_shortDelay(self):
+        """
+        Test if usb stack can handle a short delay.
+        The short delay should not cause any problems or missed packets.
+        """
+        x = 16
+        buf_out = array.array('b')
+        for i in range(0,x) : buf_out.append(48 + (x+i) % 10)
+        self.dev.ctrl_transfer(self.req_out, GZ_REQ_WRITE_LOOPBACK_BUFFER, 0, 0, buf_out)
+        delaycmd = struct.pack('<BL', GZ_INTR_CMD_DELAY, 150 * DELAY_MSEC)
+        self.dev.write(self.intr_out, delaycmd)
+        buf_in = self.dev.ctrl_transfer(self.req, GZ_REQ_READ_LOOPBACK_BUFFER, 0, 0, x)
+        self.assertEqual(len(buf_in), x,  "Should have read as much as we asked for")
+        self.assertEqual(buf_in, buf_out, "Buffers don't match!\n")
+
+    @unittest.skip("Long delay test only on demand (comment this line!)")
+    def test_longDelay(self):
+        """
+        Test if usb stack recovers correctly after a timeout.
+        The long delay will cause a timeout on the next ctrl_transfer
+        and we check that this gives an exception.  The next ctrl_transfer
+        should go through without problems.
+        """
+        x = 16
+        buf_out = array.array('b')
+        for i in range(0,x) : buf_out.append(48 + (x+i) % 10)
+        self.dev.ctrl_transfer(self.req_out, GZ_REQ_WRITE_LOOPBACK_BUFFER, 0, 0, buf_out)
+        delaycmd = struct.pack('<BL', GZ_INTR_CMD_DELAY, 1500 * DELAY_MSEC)
+        self.dev.write(self.intr_out, delaycmd)
+        # this is expected to timeout
+        self.assertRaises(usb.USBError, self.dev.ctrl_transfer, self.req, GZ_REQ_READ_LOOPBACK_BUFFER, 0, 0, x)
+        time.sleep(2)
```

Comment:
Yeah, I wanted to be safe.  Something like 500 ms may be enough here (1500 ms delay minus 1 second waiting for timeout).